README.rst
ZAP CLI
=======

.. image:: https://travis-ci.org/Grunny/zap-cli.svg?branch=master
    :target: https://travis-ci.org/Grunny/zap-cli

A commandline tool that wraps the OWASP ZAP API for controlling ZAP and
executing quick, targeted attacks.

Installation
============

To install the latest release from PyPI, you can run the following command:

::

    pip install --upgrade zapcli

To install the latest development version of ZAP CLI, you can run the
following:

::

    pip install --upgrade git+https://github.com/Grunny/zap-cli.git

To install ZAP CLI for development, including the dependencies needed
in order to run unit tests, clone this repository and use
``pip install -e .[dev]``.

Usage
=====

To use ZAP CLI, you need to set the port ZAP runs on (defaults to 8090) and
the path to the folder in which ZAP is installed. These can be set either as
commandline parameters or with the environment variables ``ZAP_PORT`` and
``ZAP_PATH``. If you have an API key set for ZAP, this can likewise be set
either as a commandline parameter or with the ``ZAP_API_KEY`` environment
variable.

ZAP CLI can then be used with the following commands:

::

    Usage: zap-cli [OPTIONS] COMMAND [ARGS]...

      ZAP CLI - A simple commandline tool for OWASP ZAP.

    Options:
      --boring            Remove color from console output.
      -v, --verbose       Add more verbose debugging output.
      --zap-path TEXT     Path to the ZAP daemon. Defaults to /zap or the value of
                          the environment variable ZAP_PATH.
      -p, --port INTEGER  Port of the ZAP proxy. Defaults to 8090 or the value of
                          the environment variable ZAP_PORT.
      --zap-url TEXT      The URL of the ZAP proxy. Defaults to http://127.0.0.1
                          or the value of the environment variable ZAP_URL.
      --api-key TEXT      The API key for using the ZAP API if required. Defaults
                          to the value of the environment variable ZAP_API_KEY.
      --log-path TEXT     Path to the directory in which to save the ZAP output
                          log file. Defaults to the value of the environment
                          variable ZAP_LOG_PATH and uses the value of --zap-path
                          if it is not set.
      --help              Show this message and exit.

    Commands:
      active-scan  Run an Active Scan.
      ajax-spider  Run the AJAX Spider against a URL.
      alerts       Show alerts at the given alert level.
      context      Manage contexts for the current session.
      exclude      Exclude a pattern from all scanners.
      open-url     Open a URL using the ZAP proxy.
      policies     Enable or list a set of policies.
      quick-scan   Run a quick scan.
      report       Generate XML, MD or HTML report.
      scanners     Enable, disable, or list a set of scanners.
      scripts      Manage scripts.
      session      Manage sessions.
      shutdown     Shutdown the ZAP daemon.
      spider       Run the spider against a URL.
      start        Start the ZAP daemon.
      status       Check if ZAP is running.

You can use ``--help`` with any of the subcommands to get information on how to use
them.

Getting started running a scan
------------------------------
In order to run a scan, you can use either the ``active-scan`` or the ``quick-scan``
command. The ``active-scan`` only runs an active scan against a URL that is already
in ZAP's site tree (i.e. has already been opened using the ``open-url`` command or
found by running the ``spider``). The ``quick-scan`` command is intended to be a way
to run quick scans of a site with most options contained within a single command
(including being able to start and shutdown ZAP before and after), so you can do
everything in one go. Without any other options passed to the command, ``quick-scan``
will open the URL to make sure it's in the site tree, run an active scan, and will
output any found alerts.

As an example, to run a quick scan of a URL that will open and spider the URL, scan
recursively, exclude URLs matching a given regex, and only use XSS and SQLi scanners,
you could run:

::

    $ zap-cli quick-scan -s xss,sqli --spider -r -e "some_regex_pattern" http://127.0.0.1/
    [INFO]            Running a quick scan for http://127.0.0.1/
    [INFO]            Issues found: 1
    +----------------------------------+--------+----------+---------------------------------------------------------------------------------+
    | Alert                            | Risk   |   CWE ID | URL                                                                             |
    +==================================+========+==========+=================================================================================+
    | Cross Site Scripting (Reflected) | High   |       79 | http://127.0.0.1/index.php?foo=%22%3E%3Cscript%3Ealert%281%29%3B%3C%2Fscript%3E |
    +----------------------------------+--------+----------+---------------------------------------------------------------------------------+

The above example is equivalent to running the following commands in order:

::

    $ zap-cli open-url http://127.0.0.1/
    [INFO]            Accessing URL http://127.0.0.1/
    $ zap-cli exclude "some_regex_pattern"
    $ zap-cli spider http://127.0.0.1/
    [INFO]            Running spider...
    $ zap-cli active-scan --scanners xss,sqli --recursive http://127.0.0.1/
    [INFO]            Running an active scan...
    $ zap-cli alerts
    [INFO]            Issues found: 1
    +----------------------------------+--------+----------+---------------------------------------------------------------------------------+
    | Alert                            | Risk   |   CWE ID | URL                                                                             |
    +==================================+========+==========+=================================================================================+
    | Cross Site Scripting (Reflected) | High   |       79 | http://127.0.0.1/index.php?foo=%22%3E%3Cscript%3Ealert%281%29%3B%3C%2Fscript%3E |
    +----------------------------------+--------+----------+---------------------------------------------------------------------------------+

The ``quick-scan`` command also has a ``--self-contained`` option (or ``-sc`` for short)
which will first try to start ZAP if it isn't running already and shutdown ZAP once the
scan is finished. For example:

::

    $ zap-cli quick-scan --self-contained --spider -r -s xss http://127.0.0.1/
    [INFO]            Starting ZAP daemon
    [INFO]            Running a quick scan for http://127.0.0.1/
    [INFO]            Issues found: 1
    +----------------------------------+--------+----------+---------------------------------------------------------------------------------+
    | Alert                            | Risk   |   CWE ID | URL                                                                             |
    +==================================+========+==========+=================================================================================+
    | Cross Site Scripting (Reflected) | High   |       79 | http://127.0.0.1/index.php?foo=%22%3E%3Cscript%3Ealert%281%29%3B%3C%2Fscript%3E |
    +----------------------------------+--------+----------+---------------------------------------------------------------------------------+
    [INFO]            Shutting down ZAP daemon

Extra start options
-------------------
You can also pass extra options to the start command of ZAP using ``--start-options`` or ``-o``
with commands that allow it. For example, to start ZAP with a custom API key you could use:

::

    $ zap-cli start --start-options '-config api.key=12345'

Or to run a self-contained quick scan (that will start ZAP and shut it down after the scan
is complete) with a custom API key, you could use:

::

    $ zap-cli --api-key 12345 quick-scan --self-contained -o '-config api.key=12345' -s xss http://127.0.0.1/

Or to run the same scan with the API key disabled:

::

    $ zap-cli quick-scan -sc -o '-config api.disablekey=true' -s xss http://127.0.0.1/

Running scans as authenticated users
------------------------------------
In order to run a scan as an authenticated user, first configure the authentication method and users for
a context using the ZAP UI (see the `ZAP help page <https://github.com/zaproxy/zap-core-help/wiki/HelpStartConceptsAuthentication>`_
for more information). Once the authentication method and users are prepared, you can then export the context
with the configured authentication method so it can be imported and used to run authenticated scans with ZAP CLI.

You can export a context with the authentication method and users configured either through the ZAP UI or using the
``context export`` ZAP CLI command. For example, to export a context with the name DevTest to a file, you could run:

::

    $ zap-cli context export --name DevTest --file-path /home/user/DevTest.context

To import the saved context for use with ZAP CLI later, you could run:

::

    $ zap-cli context import /home/user/DevTest.context

After importing the context with the configured authentication method and users, you can then provide the context name
and user name to the ``spider``, ``active-scan``, and ``quick-scan`` commands to run the scans while authenticated as
the given user. For example:

::

    $ zap-cli context import /home/user/DevTest.context
    $ zap-cli open-url "http://localhost/"
    $ zap-cli spider --context-name DevTest --user-name SomeUser "http://localhost"
    $ zap-cli active-scan --recursive -c DevTest -u SomeUser "http://localhost"
    $ zap-cli quick-scan --recursive --spider -c DevTest -u SomeUser "http://localhost"


HISTORY.rst
Release History
===============

v0.10.0 (2018-12-31)
--------------------
* Change exit codes for alerts and quick-scan commands (#29)
* Add ability to change the directory used for the log file (#62)
* Add support for outputting markdown reports
* Add ability to set the strength and threshold for policies and scanners
* Upgrade python-owasp-zap-v2.4 dependency to 0.0.14

v0.9.0 (2017-11-01)
-------------------
* Add commands to manage contexts as well as options to run the spider,
  active scan, and quick scan while authenticated as a user. (#7)

v0.8.1 (2017-07-20)
-------------------
* Upgrade python-owasp-zap-v2.4 dependency to 0.0.11

v0.8.0 (2017-07-11)
-------------------
* Fix support for unicode characters in reports
* Add JSON output format to quick-scan

v0.7.0 (2017-05-14)
-------------------
* Update zap-cli to support Python 3.5

v0.6.0 (2017-03-29)
-------------------
* Update zap-cli to support ZAP 2.6.0

v0.5.0 (2017-03-27)
-------------------
* Add commands for managing scripts

v0.4.0 (2016-10-09)
-------------------
* Add a report command to save a HTML or XML report

v0.3.0 (2016-08-28)
-------------------
* Add a status command to check if ZAP is running (#14)
* Raise an error when the ZAP executable is not found (#11)
* Upgrade python-owasp-zap-v2.4 dependency to 0.0.8

v0.2.1 (2016-05-09)
-------------------
* Handle errors when running the Spider (#9)
* Make ZAP path default to /zap if neither the environment variable nor the
  parameter are set

v0.2.0 (2016-02-21)
-------------------
* Add support for running AJAX Spider both on its own and as part of a
  quick scan.
* Add documentation to clarify the difference between active-scan and
  quick-scan, and add a few more examples of how they can work.
* Better active-scan error handling when a URL is not found in the site tree.
* Upgrade python-owasp-zap-v2.4 dependency to 0.0.7

v0.1.1 (2015-10-14)
-------------------
* Upgrade python-owasp-zap-v2.4 dependency to 0.0.5

v0.1.0 (2015-10-14)
-------------------
* Initial release of zap-cli for PyPI


zapcli/cli.py
"""
ZAP CLI.
"""

import sys

import click

from zapcli import __version__
from zapcli import helpers
from zapcli.commands.context import context_group
from zapcli.commands.policies import policies_group
from zapcli.commands.scanners import scanner_group
from zapcli.commands.scripts import scripts_group
from zapcli.commands.session import session_group
from zapcli.log import console
from zapcli.zap_helper import ZAPHelper


@click.group(help='ZAP CLI v{0} - A simple commandline tool for OWASP ZAP.'.format(__version__))
@click.option('--boring', is_flag=True, default=False, help='Remove color from console output.')
@click.option('--verbose', '-v', is_flag=True, default=False, type=bool,
              help='Add more verbose debugging output.')
@click.option('--zap-path', default='/zap', envvar='ZAP_PATH', type=str,
              help='Path to the ZAP daemon. Defaults to /zap or the value of the environment variable ZAP_PATH.')
@click.option('--port', '-p', default=8090, envvar='ZAP_PORT', type=int,
              help='Port of the ZAP proxy. Defaults to 8090 or the value of the environment variable ZAP_PORT.')
@click.option('--zap-url', default='http://127.0.0.1', envvar='ZAP_URL', type=str,
              help='The URL of the ZAP proxy. Defaults to http://127.0.0.1 or the value of the environment ' +
              'variable ZAP_URL.')
@click.option('--api-key', default='', envvar='ZAP_API_KEY', type=str,
              help='The API key for using the ZAP API if required. Defaults to the value of the environment ' +
              'variable ZAP_API_KEY.')
@click.option('--log-path', envvar='ZAP_LOG_PATH', type=str,
              help='Path to the directory in which to save the ZAP output log file. Defaults to the value of ' +
              'the environment variable ZAP_LOG_PATH and uses the value of --zap-path if it is not set.')
@click.pass_context
def cli(ctx, boring, verbose, zap_path, port, zap_url, api_key, log_path):
    """Main command line entry point."""
    console.colorize = not boring

    if verbose:
        console.setLevel('DEBUG')
    else:
        console.setLevel('INFO')

    ctx.obj = ZAPHelper(zap_path=zap_path, port=port, url=zap_url, api_key=api_key, log_path=log_path)


@cli.command('start', short_help='Start the ZAP daemon.')
@click.option('--start-options', '-o', type=str,
              help='Extra options to pass to the ZAP start command, e.g. "-config api.key=12345"')
@click.pass_obj
def start_zap_daemon(zap_helper, start_options):
    """Helper to start the daemon using the current config."""
    console.info('Starting ZAP daemon')
    with helpers.zap_error_handler():
        zap_helper.start(options=start_options)


@cli.command('shutdown')
@click.pass_obj
def shutdown_zap_daemon(zap_helper):
    """Shutdown the ZAP daemon."""
    console.info('Shutting down ZAP daemon')
    with helpers.zap_error_handler():
        zap_helper.shutdown()


@cli.command('status', short_help='Check if ZAP is running.')
@click.option('--timeout', '-t', type=int,
              help='Wait this number of seconds for ZAP to have started')
@click.pass_obj
def check_status(zap_helper, timeout):
    """
    Check if ZAP is running and able to receive API calls.

    You can provide a timeout option which is the amount of time in seconds
    the command should wait for ZAP to start if it is not currently running.
    This is useful to run before calling other commands if ZAP was started
    outside of zap-cli. For example:

        zap-cli status -t 60 && zap-cli open-url "http://127.0.0.1/"

    Exits with code 1 if ZAP is either not running or the command timed out
    waiting for ZAP to start.
    """
    with helpers.zap_error_handler():
        if zap_helper.is_running():
            console.info('ZAP is running')
        elif timeout is not None:
            zap_helper.wait_for_zap(timeout)
            console.info('ZAP is running')
        else:
            console.error('ZAP is not running')
            sys.exit(2)


@cli.command('open-url')
@click.argument('url')
@click.pass_obj
def open_url(zap_helper, url):
    """Open a URL using the ZAP proxy."""
    console.info('Accessing URL {0}'.format(url))
    zap_helper.open_url(url)


@cli.command('spider')
@click.argument('url')
@click.option('--context-name', '-c', type=str, help='Context to use if provided.')
@click.option('--user-name', '-u', type=str,
              help='Run scan as this user if provided. If this option is used, the context parameter must also ' +
              'be provided.')
@click.pass_obj
def spider_url(zap_helper, url, context_name, user_name):
    """Run the spider against a URL."""
    console.info('Running spider...')
    with helpers.zap_error_handler():
        zap_helper.run_spider(url, context_name, user_name)


@cli.command('ajax-spider')
@click.argument('url')
@click.pass_obj
def ajax_spider_url(zap_helper, url):
    """Run the AJAX Spider against a URL."""
    console.info('Running AJAX Spider...')
    zap_helper.run_ajax_spider(url)


@cli.command('active-scan', short_help='Run an Active Scan.')
@click.argument('url')
@click.option('--scanners', '-s', type=str, callback=helpers.validate_scanner_list,
              help='Comma separated list of scanner IDs and/or groups to use in the scan. Use the scanners ' +
              'subcommand to get a list of IDs. Available groups are: {0}.'.format(
                  ', '.join(['all'] + list(ZAPHelper.scanner_group_map.keys()))))
@click.option('--recursive', '-r', is_flag=True, default=False, help='Make scan recursive.')
@click.option('--context-name', '-c', type=str, help='Context to use if provided.')
@click.option('--user-name', '-u', type=str,
              help='Run scan as this user if provided. If this option is used, the context parameter must also ' +
              'be provided.')
@click.pass_obj
def active_scan(zap_helper, url, scanners, recursive, context_name, user_name):
    """
    Run an Active Scan against a URL.

    The URL to be scanned must be in ZAP's site tree, i.e. it should have already
    been opened using the open-url command or found by running the spider command.
    """
    console.info('Running an active scan...')

    with helpers.zap_error_handler():
        if scanners:
            zap_helper.set_enabled_scanners(scanners)

        zap_helper.run_active_scan(url, recursive, context_name, user_name)


@cli.command('alerts')
@click.option('--alert-level', '-l', default='High', type=click.Choice(ZAPHelper.alert_levels.keys()),
              help='Minimum alert level to include in report (default: High).')
@click.option('--output-format', '-f', default='table', type=click.Choice(['table', 'json']),
              help='Output format to print the alerts.')
@click.option('--exit-code', default=True, type=bool,
              help='Whether to set a non-zero exit code when there are any alerts of the specified ' +
              'level (default: True).')
@click.pass_obj
def show_alerts(zap_helper, alert_level, output_format, exit_code):
    """Show alerts at the given alert level."""
    alerts = zap_helper.alerts(alert_level)

    helpers.report_alerts(alerts, output_format)

    if exit_code:
        code = 1 if len(alerts) > 0 else 0
        sys.exit(code)


@cli.command('quick-scan', short_help='Run a quick scan.')
@click.argument('url')
@click.option('--self-contained', '-sc', is_flag=True, default=False,
              help='Make the scan self-contained, i.e. start the daemon, open the URL, scan it, ' +
              'and shutdown the daemon when done.')
@click.option('--scanners', '-s', type=str, callback=helpers.validate_scanner_list,
              help='Comma separated list of scanner IDs and/or groups to use in the scan. Use the scanners ' +
              'subcommand to get a list of IDs. Available groups are: {0}.'.format(
                  ', '.join(['all'] + list(ZAPHelper.scanner_group_map.keys()))))
@click.option('--spider', is_flag=True, default=False, help='If set, run the spider before running the scan.')
@click.option('--ajax-spider', is_flag=True, default=False, help='If set, run the AJAX Spider before running the scan.')
@click.option('--recursive', '-r', is_flag=True, default=False, help='Make scan recursive.')
@click.option('--alert-level', '-l', default='High', type=click.Choice(ZAPHelper.alert_levels.keys()),
              help='Minimum alert level to include in report.')
@click.option('--exclude', '-e', type=str, callback=helpers.validate_regex,
              help='Regex to exclude from all aspects of the scan')
@click.option('--start-options', '-o', type=str,
              help='Extra options to pass to the ZAP start command when the --self-contained option is used, ' +
              ' e.g. "-config api.key=12345"')
@click.option('--output-format', '-f', default='table', type=click.Choice(['table', 'json']),
              help='Output format to print the alerts.')
@click.option('--context-name', '-c', type=str, help='Context to use if provided.')
@click.option('--user-name', '-u', type=str,
              help='Run scan as this user if provided. If this option is used, the context parameter must also ' +
              'be provided.')
@click.pass_obj
def quick_scan(zap_helper, url, **options):
    """
    Run a quick scan of a site by opening a URL, optionally spidering the URL,
    running an Active Scan, and reporting any issues found.

    This command contains most scan options as parameters, so you can do
    everything in one go.

    If any alerts are found for the given alert level, this command will exit
    with a status code of 1.
    """
    if options['self_contained']:
        console.info('Starting ZAP daemon')
        with helpers.zap_error_handler():
            zap_helper.start(options['start_options'])

    console.info('Running a quick scan for {0}'.format(url))

    with helpers.zap_error_handler():
        if options['scanners']:
            zap_helper.set_enabled_scanners(options['scanners'])

        if options['exclude']:
            zap_helper.exclude_from_all(options['exclude'])

        zap_helper.open_url(url)

        if options['spider']:
            zap_helper.run_spider(url, options['context_name'], options['user_name'])

        if options['ajax_spider']:
            zap_helper.run_ajax_spider(url)

        zap_helper.run_active_scan(url, options['recursive'], options['context_name'], options['user_name'])

    alerts = zap_helper.alerts(options['alert_level'])

    helpers.report_alerts(alerts, options['output_format'])

    if options['self_contained']:
        console.info('Shutting down ZAP daemon')
        with helpers.zap_error_handler():
            zap_helper.shutdown()

    exit_code = 1 if len(alerts) > 0 else 0
    sys.exit(exit_code)


@cli.command('exclude', short_help='Exclude a pattern from all scanners.')
@click.argument('pattern', callback=helpers.validate_regex)
@click.pass_obj
def exclude_from_scanners(zap_helper, pattern):
    """Exclude a pattern from proxy, spider and active scanner."""
    with helpers.zap_error_handler():
        zap_helper.exclude_from_all(pattern)


@cli.command('report')
@click.option('--output', '-o', help='Output file for report.')
@click.option('--output-format', '-f', default='xml', type=click.Choice(['xml', 'html', 'md']),
              help='Report format.')
@click.pass_obj
def report(zap_helper, output, output_format):
    """Generate XML, MD or HTML report."""
    if output_format == 'html':
        zap_helper.html_report(output)
    elif output_format == 'md':
        zap_helper.md_report(output)
    else:
        zap_helper.xml_report(output)

    console.info('Report saved to "{0}"'.format(output))


# Add subcommand groups
cli.add_command(context_group)
cli.add_command(policies_group)
cli.add_command(scanner_group)
cli.add_command(scripts_group)
cli.add_command(session_group)
